Author: Guy, Michael, Joshua
Summary: This is a comprenhensive simulation engine for ambulance simulation. It uses the solution from two-stage stochastic program and robust programs to simulate how ambulances would perform. The simulation mostly runs using a priority-queue of events with respect to time. Events include things like ambulance arrives to the scene of call, ambulance arrives to hospital, ect.

In [14]:
import Pkg #this line doesn't seem to run either way for me
Pkg.add("Pkg")
Pkg.add("Dates")

  Resolving package versions...
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Project.toml`
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Manifest.toml`
  Resolving package versions...
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Project.toml`
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Manifest.toml`


In [15]:
"""This activates packages, namely ambulance deployment"""

guypath = "/Users/guyfarmer/ambulance_connect/AmbulanceDeployement/AmbulanceDeployment.jl-legacy/"
michaelpath= "/home/mhilborn/Desktop/UT-Spring2021/SeniorLab/AmbulanceDeployement/AmbulanceDeployment.jl-legacy"
joshpath = "C:/Users/Owner/Documents/Austin/AmbulanceDeployment/AmbulanceDeployment.jl-legacy/"

currentpath = joshpath

cd(currentpath)
Pkg.activate(".")

 Activating environment at `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Project.toml`


Warning: Package AmbulanceDeployment does not have Random in its dependencies:
│ - If you have AmbulanceDeployment checked out for development and have
│   added Random as a dependency but haven't updated your primary
│   environment's manifest file, try `Pkg.resolve()`.
│ - Otherwise you may need to report an issue with AmbulanceDeployment
└ Loading Random into AmbulanceDeployment from project dependency, future warnings for AmbulanceDeployment are suppressed.

In [16]:
using AmbulanceDeployment

In [17]:
using DataFrames, JLD, Distributions, CSV, Random, JSON, DelimitedFiles, Dates #Plots I don't think we use Plots
import DataStructures: PriorityQueue, enqueue!, dequeue!
include(".//src//problem.jl")
include(".//src//model.jl")

In [18]:
"""Helper functions for the simuation engine?"""
#=
Author : Ng Yeesian
Modified : Guy Farmer / Zander Tedjo / Will Worthington / Michael Hilborn
generates the closest dispatch model
Summary :
Input :
Output :
=#
mutable struct ClosestDispatch <: DispatchModel
    drivetime::DataFrame
    candidates::Vector{Vector{Int}}
    available::Vector{Int}
    assignment::Vector{Int} # which location the ambulance is assigned to
    ambulances::Vector{Vector{Int}} # list of ambulances assigned to each location
    status::Vector{Symbol} # the current status of the ambulance
        # possible statuses: :available, :responding, :atscene, :conveying, :returning
    fromtime::Vector{Int} # the time it started the new status
    hospital::Vector{Int}
end

#=
Summary :
Input :
Output : ClosestDispatch struct
=#
function ClosestDispatch(p::DeploymentProblem, drivetime::DataFrame,distribution::Vector{Int})
    candidates = Any[]
    I = 1:p.nlocations
    for region in 1:p.nregions
        push!(candidates, I[vec(p.coverage[region,:])])
    end
    #println("creating the dispatch problem :$distribution")

    assignment = zeros(Int, p.nambulances)
    ambulances = [Int[] for i in 1:p.nlocations]
    k = 1
    for i in eachindex(distribution), j in 1:distribution[i]
        assignment[k] = i
        push!(ambulances[i], k)
        k += 1
    end
    status = fill(:available, p.nambulances)
    fromtime = zeros(Int, p.nambulances)
    hospital = zeros(Int, p.nambulances)
    
    ClosestDispatch(drivetime, candidates, distribution, assignment, ambulances, status, fromtime, hospital)

end

function ClosestDispatch(p::DeploymentProblem, problem::DispatchProblem)
    stn_names = [Symbol("stn$(i)_min") for i in 1:size(p.coverage,2)]
    ClosestDispatch(p, problem.emergency_calls[:, stn_names], problem.available)
end

function update_ambulances!(model::ClosestDispatch, i::Int, delta::Int)
    model.available[i] += delta
    @assert model.available[i] >= 0
end

function closest_available(dispatch::ClosestDispatch, id::Int, problem::DispatchProblem)
    location = 0
    min_time = Inf
    for i in dispatch.candidates[problem.emergency_calls[id, :neighborhood]]
        if problem.available[i] > 0 && dispatch.drivetime[id, i] < min_time
            #println("apparently this station is available : $i and here is the amount of ambulances avalable: $(problem.available[i])")
            location = i
            min_time = dispatch.drivetime[id, i]
        end
    end
    location
end

function respond_to!(dispatch::ClosestDispatch, i::Int, t::Int)
    @assert length(dispatch.ambulances[i]) > 0 "$i: $(dispatch.ambulances[i])"
    amb = popfirst!(dispatch.ambulances[i])
    # @assert redeploy.hospital[amb] == 0
    @assert amb != 0
    @assert dispatch.status[amb] == :available "$amb: $(dispatch.status[amb])"
    dispatch.status[amb] = :responding
    #dispatch.fromtime[amb] = t
    amb
end

function arriveatscene!(dispatch::ClosestDispatch, amb::Int, t::Int)
    @assert dispatch.status[amb] == :responding "$amb: $(dispatch.status[amb])"
    @assert dispatch.hospital[amb] == 0
    dispatch.status[amb] = :atscene
    dispatch.fromtime[amb] = t
end

function arriveathospital!(dispatch::DispatchModel, amb::Int, hosp::Int, t::Int)
    @assert dispatch.status[amb] == :atscene "$amb: $(dispatch.status[amb])"
    @assert dispatch.hospital[amb] != 0
    dispatch.status[amb] = :atHospital
    dispatch.fromtime[amb] = t
    dispatch.hospital[amb] = hosp
end

function returning_to!(dispatch::DispatchModel, amb::Int, t::Int)
    @assert dispatch.status[amb] == :atHospital "$amb: $(dispatch.status[amb])"
    @assert dispatch.hospital[amb] != 0
    dispatch.status[amb] = :returning
    dispatch.fromtime[amb] = t
    dispatch.assignment[amb]
end

returning_to! (generic function with 1 method)

In [50]:
""" simulation parameters """
turnaround = Distributions.LogNormal(3.65, 0.3)#Distributions.LogNormal(3.089, 0.3) #Distributions.LogNormal(3.65, 0.3)
lambda = 0
local_path = ""
model_name = "Stochastic"
namb = 40
ncalls = 1000
#using AmbulanceDeployment

pwd()

cd(currentpath * "/test/")
pwd()

num_regions = 3199
hospitals = CSV.File(string(local_path,"../test/austin_data_3200/hospitals.csv")) |> DataFrame
stations = CSV.File(string(local_path,"../test/austin_data_3200/stations.csv")) |> DataFrame
hourly_calls = CSV.File("../test/austin_data_3200/WeekdayCalls.csv") |> DataFrame
adjacent_nbhd = CSV.File("../test/austin_data_3200/adjacent_nbhd.csv") |> DataFrame
test_calls = CSV.File("../test/austin_data_3200/austin_test_calls_v3.csv")|> DataFrame 
coverage = CSV.read("../test/austin_data_3200/coverage_regression.csv", DataFrame, header=false) #coverage_real is austin-data
coverage = delete!(coverage,1) #trim df header (1st row)
coverage = select(coverage,2:3234) #trim df header (1st column)
coverage = Matrix(coverage)
#coverage = convert(Array{Bool, 2}, coverage)
coverage = convert(Array{Bool,2},coverage')

# num_regions = 200
# hospitals = CSV.File(string(local_path,"../test/austin_data/hospitals.csv")) |> DataFrame
# stations = CSV.File(string(local_path,"../test/austin_data/stations.csv")) |> DataFrame
# hourly_calls = CSV.File("../test/austin_data/WeekdayCalls.csv") |> DataFrame
# adjacent_nbhd = CSV.File("../test/austin_data/adjacent_nbhd.csv") |> DataFrame
#CALLS IN MINUTES #test_calls = CSV.File("../test/austin_data/austin_test_calls_v1.csv")|> DataFrame  
# coverage = CSV.read("../test/austin_data/coverage.csv", DataFrame, header=false) #coverage_real is austin-data
# coverage = Matrix(coverage)
# coverage = convert(Array{Bool, 2}, coverage)

solverstats = JLD.load("../src/outputs/austin_team_stats_1_03.jld")
amb_deployment = solverstats["amb_deployment"]
model_dict = Dict{String, Symbol}("Stochastic"=>:Stochastic, "Robust01"=>:Robust01, "Robust005"=>:Robust005, "Robust001"=>:Robust001, "Robust0001"=>:Robust0001,
          "Robust00001"=>:Robust00001, "MEXCLP"=>:MEXCLP, "MALP"=>:MALP)


include("..//src//problem.jl")

p = DeploymentProblem(
          num_regions,
          hourly_calls,
          adjacent_nbhd,
          coverage,
          namb = namb,
          train_filter = (hourly_calls[!,:year] .== 2019) .* (hourly_calls[!,:month] .<= 3)
          )

print("n ambulances ",p.nambulances)
print("\n locations  ",p.nlocations)
print("\n nregions ",p.nregions)
# print("\n train ", p.train)
# print("\n test ", p.test)
print("\n coverage ",size(p.coverage))
print("\n demand ",size(p.demand))
print("\n adjacency ",size(p.adjacency))

#update 1: remove negative indices
test_calls = test_calls[1:ncalls,:]
println("running $namb ambuances & $ncalls calls")

# distribution of ambulances across all of the stations
x = amb_deployment[model_dict[model_name]][namb]#convert(Array{Int,1},ones(44))#amb_deployment[model_dict[model_name]][namb]

#establishes the dispatch problem
problem = DispatchProblem(test_calls, hospitals, stations, p.coverage, x, turnaround=turnaround)

#establishes the dispatch model
dispatch = ClosestDispatch(p, problem)
Random.seed!(1234)

40
44
3199
(0, 2)
(0,)
(0,)
(3233, 44)
Array{Bool,2}
(3199, 3199)
BitArray{2}
n ambulances 40
 locations  44
 nregions 3199
 coverage (3233, 44)
 demand (0, 2)
 adjacency (3199, 3199)running 40 ambuances & 1000 calls


MersenneTwister(UInt32[0x000004d2], Random.DSFMT.DSFMT_state(Int32[-1393240018, 1073611148, 45497681, 1072875908, 436273599, 1073674613, -2043716458, 1073445557, -254908435, 1072827086  …  -599655111, 1073144102, 367655457, 1072985259, -1278750689, 1018350124, -597141475, 249849711, 382, 0]), [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], UInt128[0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000  …  0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x000000000000

###### problem.available

In [51]:
#test_calls
#problem.emergency_calls[:, Symbol("stn$(1)_min")]

In [52]:
function logRegression(gridTime)
    if(gridTime < 0)
        print("grid time less than 0: ", gridTime)
    end
    gridTime = gridTime / 60
    slope = 0.23868463365149437
    intercept = 1.261777659720721
    ax = log(gridTime)*slope
    new_time = exp(ax + intercept)
    new_time =  new_time * 60
    new_time
end

logRegression (generic function with 1 method)

In [58]:
"""Simulation Engine"""

Pkg.add("Query")
using Query

struct EMSEngine{T}
    eventlog::DataFrame
    eventqueue::PriorityQueue{T,Int,Base.Order.ForwardOrdering}
    shortqueue::PriorityQueue{T,Int,Base.Order.ForwardOrdering}
    guiArray::Array{Any, 1}
end

struct gui_event
    #call made / call responded / ambulance arrived / shortfall
    event_type::String
    #neighborhood where call is made from
    neighborhood_id::Int
    # station where call is made from
    deployment_id::Int
    # id of the event generated
    event_id::Int
    ambulance_id::Int
    arrival_time::Int
    remaining_amb::Int
    # current timestamp of event
    timestamp::DateTime
end

## generates the priority queue and instantiates the EMSEngine struct
function generateCalls(problem::DispatchProblem)
    ncalls = nrow(problem.emergency_calls)
    eventlog = DataFrame(
        id = 1:ncalls,
        dispatch_from = zeros(Int, ncalls),
        waittime = fill(0.0, ncalls),
        responsetime = fill(Inf, ncalls),
        scenetime = fill(Inf, ncalls),
        conveytime = fill(Inf, ncalls),
        returntime = fill(Inf, ncalls),
        return_to = zeros(Int, ncalls),
        return_type = fill(:station, ncalls),
        hospital = zeros(Int, ncalls),
        ambulance = zeros(Int, ncalls)
    )
    eventqueue = PriorityQueue{Tuple{Symbol,Int,Int,Int},Int}()
    shortqueue = PriorityQueue{Tuple{Symbol,Int,Int,Int},Int}()
    guiArray = Any[]
    for i in 1:nrow(problem.emergency_calls)
        t = problem.emergency_calls[i, :arrival_seconds]
        enqueue!(eventqueue, (:call, i, t, problem.emergency_calls[i, :neighborhood]), t)
    end
    EMSEngine{Tuple{Symbol,Int,Int,Int}}(eventlog, eventqueue, shortqueue, guiArray) ## returns the ems engine struct
end

function simulate_events!(problem::DispatchProblem, dispatch::ClosestDispatch)
    engine = generateCalls(problem)
    while(~isempty(engine.eventqueue))
        (event, id, t, value) = dequeue!(engine.eventqueue)

        if event == :call
            # if the event is reachable then change event to station2call
            # if not it is a shortfall
            println("calling event id: $id time: $t value: $value")
            call_event!(engine, problem, dispatch, id, t, value)
           
        elseif event == :station2call
            #once the ambulance arrives at the station change event to call2hospital
            println("Ambulance has arrived on the scene for event $id at time $t")
            station2call_event!(engine, problem, dispatch, id, t, value)
        elseif event == :call2hospital
            #once the ambulance arrives at the hospital it will hand over the patient and then 
            #change the deploy back to its station
            println("Ambulance for event $id has arrived at the hospital at time $t")
            call2hosp_event!(engine, problem, dispatch, id, t, value)

        elseif event == :hospital2station
            #once the ambulance drops off the patient it will reroute to the station
            println("Ambulance for event $id is returning to the station at $t")
            return_event!(engine, problem, dispatch, id, t, value)
            
        elseif event == :done
            #increase the number of ambulances at that station now that it is done
            done_event!(engine, problem, dispatch, id, t, value)
            
        end
    end
    println("This is the Short Queue")
    while(~isempty(engine.shortqueue))
       (event, id, t, value) = dequeue!(engine.shortqueue)
        print("$event, $id, $t, $value")
        print("This is not empty")
    end
    engine.eventlog , engine.guiArray
end


function call_event!(
        ems::EMSEngine,
        problem::DispatchProblem,
        dispatch::DispatchModel,
        id::Int, # the id of the emergency call
        t::Int, # the time of the emergency call
        nbhd::Int; # the neighborhood the call is from
    )

    #check if there is an ambulance within the coverage matrix
    if sum(problem.deployment[problem.coverage[nbhd,:]]) == 0
        @assert false "$id: no ambulance reachable for call at $nbhd"
    #check if one of the ambulances within the coverage is available
    elseif sum(problem.available[problem.coverage[nbhd,:]]) > 0
        i = closest_available(dispatch, id, problem)
        println("closest available station : $i")
        @assert i > 0 "$id: dispatch from $i to nbhd $nbhd" # assume valid i (enforced by <if> condition)
        update_ambulances!(dispatch, i, -1)
        ems.eventlog[id, :dispatch_from] = i
        @assert problem.available[i] >= 0
        
        travel_time = ceil(Int, problem.emergency_calls[id, Symbol("stn$(i)_min")]) #60*
        println(travel_time)
        #travel_time = logRegression(travel_time) # gridTime2Bin(travel_time) #adjust time
        travel_time = ceil(travel_time) #time mod
        println(travel_time)
        @assert travel_time >= 0
        if(t != problem.emergency_calls[id, :arrival_seconds])
            print("This is a shortfall")
        end
        ems.eventlog[id, :responsetime] = (travel_time + t - problem.emergency_calls[id,:arrival_seconds]) / 60 # minutes

        amb = respond_to!(dispatch, i, t)
        ems.eventlog[id, :ambulance] = amb

        event = gui_event("call responded", problem.emergency_calls[id, :neighborhood], i, id, amb,-1, problem.available[i] ,Dates.now())
        num_ambulances_array = Array{Integer}(undef, size(problem.available, 1))
        for i in 1:size(problem.available, 1)
            num_ambulances_array[i] = problem.available[i]
        end
        push!(ems.guiArray,event)
        push!(ems.guiArray,num_ambulances_array)
        enqueue!(ems.eventqueue, (:station2call, id, t + travel_time, amb), t + travel_time)
    #else queue it
    else
        # println(id, ": call from ", nbhd, " queued behind ", problem.wait_queue[nbhd])
        # event = gui_event("call made", nbhd, -1, id, -1,-1,-1,Dates.now())
        # push!(ems.guiArray,event)
        # problem.shortfalls = problem.shortfalls + 1
        # #push!(problem.shortfalls) # count shortfalls
        # push!(problem.wait_queue[nbhd], id) # queue the emergency call
        event = gui_event("shortfall", problem.emergency_calls[id, :neighborhood], -1, id, -1,-1,-1 ,Dates.now())
        push!(ems.guiArray, event)
        enqueue!(ems.shortqueue, (:call, id, t, problem.emergency_calls[id, :neighborhood]), t)
        println(id, ": call from ", nbhd, " is a shortfall")
        println("and there are this many amb available: ", sum(problem.available))
    end
end

function station2call_event!(
        ems::EMSEngine,
        problem::DispatchProblem,
        dispatch::DispatchModel,
        id::Int, # the id of the emergency call
        t::Int, # the time of the emergency call
        amb::Int
    )
    arriveatscene!(dispatch, amb, t)
    # time the ambulance spends at the scene
    scene_time = ceil(Int,60*0.4*rand(problem.turnaround)) # 60sec*0.4*mean(40minutes) ~ 15minutes
    ems.eventlog[id, :scenetime] = scene_time / 60 # minutes
    @assert scene_time > 0
    event = gui_event("ambulance arrived", problem.emergency_calls[id, :neighborhood], -1, id, amb, t,-1 ,Dates.now())
    push!(ems.guiArray, event)
    enqueue!(ems.eventqueue, (:call2hospital, id, t + scene_time, amb), t + scene_time)
end

function call2hosp_event!(
        ems::EMSEngine,
        problem::DispatchProblem,
        dispatch::DispatchModel,
        id::Int, # the id of the emergency call
        t::Int, # the time of the emergency call
        amb::Int
    )
    h = let mintime = Inf, minindex = 0
        for h in 1:nrow(problem.hospitals)
            traveltime = problem.emergency_calls[id, Symbol("hosp$(h)_min")]
            if !isna(traveltime) && !(traveltime=="NA") && traveltime < mintime
                @assert traveltime >= 0
                minindex = h; mintime = traveltime
            end
        end
        minindex
    end
    @assert h != 0
    dispatch.hospital[amb] = ems.eventlog[id, :hospital] = h
    #60*15 + ceil(Int, 60*problem.emergency_calls[id, Symbol("hosp$(h)_min")]) # ~20minutes why the hell would we add 15 minutes and then also the hopsital min!! smh
    conveytime = ceil(Int, 60*problem.emergency_calls[id, Symbol("hosp$(h)_min")]) # ~20minutes
    conveytime = ceil(logRegression(conveytime)) #adjust time
    ems.eventlog[id, :conveytime] = conveytime / 60 # minutes
    @assert conveytime >= 0 conveytime
    arriveathospital!(dispatch, amb, h, t)
    
    #         #call made / call responded / ambulance arrived / shortfall
#     event_type::String
#     #neighborhood where call is made from
#     neighborhood_id::Int
#     # station where call is made from
#     deployment_id::Int
#     # id of the event generated
#     event_id::Int
#     ambulance_id::Int
#     arrival_time::Int
#     remaining_amb::Int
#     # current timestamp of event
#     timestamp::DateTime
    event = gui_event("ambulance hospital", problem.emergency_calls[id, :neighborhood], -1, id, amb, t,-1 ,Dates.now())
    push!(ems.guiArray, event)
    
    
    enqueue!(ems.eventqueue, (:hospital2station, id, t+conveytime, amb), t+conveytime)
end

function return_event!(
        ems::EMSEngine,
        problem::DispatchProblem,
        dispatch::DispatchModel,
        id::Int,
        t::Int,
        amb::Int
    )
    stn = returning_to!(dispatch, amb, t)
    h = dispatch.hospital[amb]
    returntime = ceil(Int,60*problem.hospitals[h, Symbol("stn$(stn)_min")]) # ~ 10minutes
    #returntime = ceil(returntime * .9) #time mod
    dispatch.hospital[amb] = 0
    ems.eventlog[id, :returntime] = returntime / 60 # minutes
    @assert returntime >= 0 returntime
    t_end = t + returntime
    
#         #call made / call responded / ambulance arrived / shortfall
#     event_type::String
#     #neighborhood where call is made from
#     neighborhood_id::Int
#     # station where call is made from
#     deployment_id::Int
#     # id of the event generated
#     event_id::Int
#     ambulance_id::Int
#     arrival_time::Int
#     remaining_amb::Int
#     # current timestamp of event
#     timestamp::DateTime
    
    event = gui_event("ambulance returning", problem.emergency_calls[id, :neighborhood], -1, id, amb, t,-1 ,Dates.now())
    push!(ems.guiArray, event)
    
    
    enqueue!(ems.eventqueue, (:done, id, t_end, amb), t_end)
    # num_ambulances_array = Array{Integer}(undef, size(problem.available, 1))
    # for i in 1:size(problem.available, 1)
    #     num_ambulances_array[i] = problem.available[i]
    # end
    # push!(ems.guiArray,num_ambulances_array)
end

function done_event!(
        ems::EMSEngine,
        problem::DispatchProblem,
        dispatch::DispatchModel,
        id::Int,
        t::Int,
        amb::Int
    )
    i = dispatch.assignment[amb]
    println("Ambulance from event $id has returned to station $i")
    @assert i > 0 "$id: dispatch from $i to nbhd $nbhd" # assume valid i (enforced by <if> condition)
    update_ambulances!(dispatch, i, 1)
    push!(dispatch.ambulances[i], amb)
    dispatch.status[amb] = :available
    ems.eventlog[id, :return_to] = id
    @assert problem.available[i] > 0
    
#         #call made / call responded / ambulance arrived / shortfall
#     event_type::String
#     #neighborhood where call is made from
#     neighborhood_id::Int
#     # station where call is made from
#     deployment_id::Int
#     # id of the event generated
#     event_id::Int
#     ambulance_id::Int
#     arrival_time::Int
#     remaining_amb::Int
#     # current timestamp of event
#     timestamp::DateTime
    
    event = gui_event("ambulance returned", problem.emergency_calls[id, :neighborhood], -1, id, amb, t,-1 ,Dates.now())
    push!(ems.guiArray, event)
    
    while(~isempty(ems.shortqueue))
        # record time difference
        (event, short_id, short_t, value) = dequeue!(ems.shortqueue)
        enqueue!(ems.eventqueue, (event, short_id, t , value), short_t)
    end
end


  Resolving package versions...
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Project.toml`
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Manifest.toml`


done_event! (generic function with 1 method)

In [59]:
"""call simulation"""
#@time df, guiArray = simulate_events!(problem, dispatch)
df, guiArray = simulate_events!(problem, dispatch)

calling event id: 1 time: 38 value: 1012
closest available station : 1
143
143
Ambulance has arrived on the scene for event 1 at time 181
calling event id: 2 time: 575 value: 220
closest available station : 31
646
646
calling event id: 3 time: 673 value: 1687
closest available station : 25
477
477
calling event id: 4 time: 691 value: 1536
closest available station : 33
328
328
Ambulance for event 1 has arrived at the hospital at time 864
Ambulance has arrived on the scene for event 4 at time 1019
Ambulance has arrived on the scene for event 3 at time 1150
calling event id: 5 time: 1166 value: 1621
closest available station : 14
375
375
calling event id: 6 time: 1187 value: 578
closest available station : 22
222
222
Ambulance has arrived on the scene for event 2 at time 1221
Ambulance has arrived on the scene for event 6 at time 1409
calling event id: 7 time: 1450 value: 1169
closest available station : 16
927
927
Ambulance has arrived on the scene for event 5 at time 1541
calling event

Ambulance from event 24 has returned to station 17
calling event id: 32 time: 8673 value: 1008
closest available station : 8
272
272
Ambulance for event 23 is returning to the station at 8723
Ambulance has arrived on the scene for event 31 at time 8819
Ambulance from event 22 has returned to station 9
Ambulance has arrived on the scene for event 32 at time 8945
Ambulance for event 26 has arrived at the hospital at time 8946
Ambulance for event 25 has arrived at the hospital at time 8947
Ambulance from event 23 has returned to station 15
calling event id: 33 time: 9154 value: 575
closest available station : 2
342
342
Ambulance for event 27 has arrived at the hospital at time 9208
Ambulance for event 30 has arrived at the hospital at time 9290
calling event id: 34 time: 9489 value: 1501
closest available station : 14
619
619
Ambulance has arrived on the scene for event 33 at time 9496
Ambulance for event 31 has arrived at the hospital at time 9564
calling event id: 35 time: 9626 value: 8

Ambulance from event 44 has returned to station 30
Ambulance for event 51 has arrived at the hospital at time 16337
Ambulance for event 53 has arrived at the hospital at time 16456
Ambulance has arrived on the scene for event 57 at time 16477
Ambulance from event 48 has returned to station 1
Ambulance from event 47 has returned to station 43
Ambulance for event 56 has arrived at the hospital at time 16936
Ambulance for event 52 is returning to the station at 16966
Ambulance for event 49 is returning to the station at 17076
Ambulance for event 54 has arrived at the hospital at time 17081
Ambulance for event 50 is returning to the station at 17137
calling event id: 58 time: 17141 value: 1316
closest available station : 1
541
541
Ambulance for event 53 is returning to the station at 17232
calling event id: 59 time: 17297 value: 1438
closest available station : 17
715
715
Ambulance from event 52 has returned to station 5
Ambulance for event 51 is returning to the station at 17343
Ambulance

closest available station : 4
254
254
Ambulance has arrived on the scene for event 83 at time 27032
Ambulance for event 76 is returning to the station at 27073
Ambulance for event 79 has arrived at the hospital at time 27137
Ambulance has arrived on the scene for event 82 at time 27210
Ambulance for event 77 is returning to the station at 27215
Ambulance has arrived on the scene for event 81 at time 27377
Ambulance from event 73 has returned to station 28
calling event id: 84 time: 27471 value: 513
closest available station : 2
93
93
Ambulance for event 78 is returning to the station at 27549
Ambulance from event 76 has returned to station 14
Ambulance has arrived on the scene for event 84 at time 27564
Ambulance has arrived on the scene for event 80 at time 27655
calling event id: 85 time: 27753 value: 1010
closest available station : 5
333
333
Ambulance for event 83 has arrived at the hospital at time 27779
Ambulance for event 82 has arrived at the hospital at time 27911
calling even

closest available station : 17
407
407
calling event id: 108 time: 38739 value: 1624
closest available station : 4
680
680
Ambulance has arrived on the scene for event 106 at time 38767
Ambulance from event 102 has returned to station 9
Ambulance has arrived on the scene for event 107 at time 39049
Ambulance from event 101 has returned to station 22
calling event id: 109 time: 39060 value: 1808
closest available station : 14
748
748
Ambulance for event 105 has arrived at the hospital at time 39182
Ambulance for event 103 is returning to the station at 39304
Ambulance for event 104 has arrived at the hospital at time 39409
Ambulance has arrived on the scene for event 108 at time 39419
Ambulance for event 106 has arrived at the hospital at time 39657
Ambulance from event 103 has returned to station 1
Ambulance has arrived on the scene for event 109 at time 39808
Ambulance for event 107 has arrived at the hospital at time 39890
Ambulance for event 105 is returning to the station at 40115


Ambulance for event 131 has arrived at the hospital at time 48321
Ambulance has arrived on the scene for event 132 at time 48334
Ambulance from event 123 has returned to station 3
Ambulance from event 120 has returned to station 17
Ambulance from event 127 has returned to station 8
Ambulance for event 128 is returning to the station at 48494
Ambulance for event 124 is returning to the station at 48540
Ambulance has arrived on the scene for event 133 at time 48566
Ambulance for event 129 is returning to the station at 48606
calling event id: 134 time: 48703 value: 1198
closest available station : 4
358
358
Ambulance for event 130 is returning to the station at 48932
calling event id: 135 time: 49029 value: 455
closest available station : 22
609
609
Ambulance from event 128 has returned to station 26
Ambulance has arrived on the scene for event 134 at time 49061
Ambulance from event 124 has returned to station 7
Ambulance for event 133 has arrived at the hospital at time 49218
Ambulance 

Ambulance for event 148 is returning to the station at 54134
Ambulance for event 156 has arrived at the hospital at time 54156
Ambulance has arrived on the scene for event 159 at time 54183
Ambulance from event 149 has returned to station 11
Ambulance for event 152 has arrived at the hospital at time 54213
calling event id: 164 time: 54324 value: 1008
closest available station : 1
626
626
Ambulance has arrived on the scene for event 160 at time 54340
Ambulance from event 148 has returned to station 34
Ambulance from event 146 has returned to station 9
Ambulance for event 153 has arrived at the hospital at time 54461
Ambulance from event 144 has returned to station 22
Ambulance has arrived on the scene for event 163 at time 54541
Ambulance has arrived on the scene for event 162 at time 54564
Ambulance for event 151 is returning to the station at 54618
Ambulance has arrived on the scene for event 161 at time 54660
Ambulance for event 154 has arrived at the hospital at time 54703
Ambulanc

closest available station : 4
602
602
Ambulance for event 180 has arrived at the hospital at time 60146
Ambulance for event 178 has arrived at the hospital at time 60159
Ambulance for event 173 has arrived at the hospital at time 60187
Ambulance for event 183 has arrived at the hospital at time 60205
calling event id: 190 time: 60240 value: 900
closest available station : 44
195
195
Ambulance from event 174 has returned to station 25
Ambulance from event 172 has returned to station 15
Ambulance has arrived on the scene for event 187 at time 60359
Ambulance for event 169 is returning to the station at 60382
Ambulance from event 175 has returned to station 5
Ambulance has arrived on the scene for event 190 at time 60435
calling event id: 191 time: 60436 value: 900
closest available station : 8
1372
1372
calling event id: 192 time: 60466 value: 1684
closest available station : 25
203
203
Ambulance has arrived on the scene for event 186 at time 60496
Ambulance for event 176 is returning to

Ambulance has arrived on the scene for event 210 at time 64982
Ambulance has arrived on the scene for event 211 at time 65075
Ambulance for event 201 has arrived at the hospital at time 65094
Ambulance for event 205 has arrived at the hospital at time 65099
Ambulance from event 200 has returned to station 26
Ambulance for event 206 has arrived at the hospital at time 65138
Ambulance from event 203 has returned to station 3
Ambulance for event 207 has arrived at the hospital at time 65315
Ambulance from event 202 has returned to station 19
Ambulance for event 209 has arrived at the hospital at time 65624
Ambulance for event 204 is returning to the station at 65679
Ambulance for event 211 has arrived at the hospital at time 65747
Ambulance for event 206 is returning to the station at 65924
Ambulance for event 205 is returning to the station at 65982
calling event id: 212 time: 66018 value: 1604
closest available station : 34
407
407
Ambulance for event 208 has arrived at the hospital at 

Ambulance for event 230 has arrived at the hospital at time 70714
Ambulance from event 225 has returned to station 17
Ambulance has arrived on the scene for event 239 at time 70808
Ambulance for event 223 is returning to the station at 70815
Ambulance for event 224 is returning to the station at 70861
calling event id: 242 time: 70887 value: 1989
closest available station : 17
606
606
Ambulance from event 221 has returned to station 28
Ambulance for event 232 has arrived at the hospital at time 70913
Ambulance for event 226 has arrived at the hospital at time 70920
Ambulance for event 228 has arrived at the hospital at time 70926
calling event id: 243 time: 70931 value: 889
closest available station : 1
259
259
Ambulance for event 235 has arrived at the hospital at time 70977
Ambulance for event 222 is returning to the station at 70992
Ambulance for event 229 is returning to the station at 71180
Ambulance has arrived on the scene for event 243 at time 71190
Ambulance from event 224 has

1231
1231
Ambulance from event 247 has returned to station 15
Ambulance for event 253 is returning to the station at 75406
Ambulance has arrived on the scene for event 263 at time 75412
Ambulance has arrived on the scene for event 267 at time 75427
Ambulance has arrived on the scene for event 261 at time 75458
Ambulance has arrived on the scene for event 264 at time 75489
Ambulance for event 250 is returning to the station at 75551
Ambulance for event 254 has arrived at the hospital at time 75576
Ambulance for event 255 has arrived at the hospital at time 75604
Ambulance has arrived on the scene for event 266 at time 75620
Ambulance has arrived on the scene for event 265 at time 75647
Ambulance for event 245 is returning to the station at 75689
Ambulance for event 260 has arrived at the hospital at time 75734
Ambulance for event 251 is returning to the station at 75740
Ambulance for event 257 is returning to the station at 75750
Ambulance from event 253 has returned to station 3
callin

Ambulance for event 279 has arrived at the hospital at time 79655
calling event id: 289 time: 79668 value: 823
closest available station : 9
311
311
Ambulance from event 274 has returned to station 18
calling event id: 290 time: 79753 value: 1559
closest available station : 17
641
641
calling event id: 291 time: 79929 value: 773
closest available station : 43
468
468
Ambulance has arrived on the scene for event 289 at time 79979
Ambulance from event 276 has returned to station 19
calling event id: 292 time: 80005 value: 2088
closest available station : 34
2482
2482
Ambulance has arrived on the scene for event 287 at time 80008
calling event id: 293 time: 80022 value: 773
closest available station : 44
589
589
Ambulance for event 280 is returning to the station at 80025
calling event id: 294 time: 80056 value: 1500
closest available station : 17
570
570
Ambulance for event 282 is returning to the station at 80097
Ambulance for event 283 is returning to the station at 80146
Ambulance for

closest available station : 34
527
527
Ambulance from event 303 has returned to station 1
Ambulance has arrived on the scene for event 313 at time 86139
Ambulance has arrived on the scene for event 309 at time 86175
Ambulance for event 307 has arrived at the hospital at time 86179
Ambulance for event 308 has arrived at the hospital at time 86205
Ambulance has arrived on the scene for event 310 at time 86279
Ambulance for event 305 has arrived at the hospital at time 86368
Ambulance from event 301 has returned to station 5
Ambulance has arrived on the scene for event 312 at time 86445
Ambulance has arrived on the scene for event 314 at time 86554
Ambulance has arrived on the scene for event 311 at time 86554
calling event id: 315 time: 86690 value: 1507
closest available station : 11
921
921
calling event id: 316 time: 86758 value: 642
closest available station : 22
446
446
calling event id: 317 time: 86769 value: 960
closest available station : 43
757
757
Ambulance for event 310 has ar

Ambulance for event 333 has arrived at the hospital at time 92831
Ambulance from event 328 has returned to station 2
Ambulance for event 332 has arrived at the hospital at time 92972
Ambulance for event 330 is returning to the station at 93030
Ambulance for event 331 is returning to the station at 93134
Ambulance for event 338 has arrived at the hospital at time 93144
calling event id: 339 time: 93160 value: 1629
closest available station : 25
865
865
Ambulance from event 330 has returned to station 8
Ambulance for event 326 is returning to the station at 93216
Ambulance for event 335 is returning to the station at 93338
Ambulance for event 334 has arrived at the hospital at time 93434
Ambulance for event 336 has arrived at the hospital at time 93436
Ambulance for event 337 has arrived at the hospital at time 93486
Ambulance from event 335 has returned to station 8
Ambulance from event 331 has returned to station 3
calling event id: 340 time: 93601 value: 1927
closest available station

Ambulance for event 358 has arrived at the hospital at time 101035
Ambulance has arrived on the scene for event 362 at time 101136
Ambulance from event 357 has returned to station 11
Ambulance for event 359 is returning to the station at 101399
Ambulance for event 361 has arrived at the hospital at time 101491
calling event id: 363 time: 101497 value: 1542
closest available station : 34
272
272
Ambulance for event 362 has arrived at the hospital at time 101512
calling event id: 364 time: 101634 value: 1071
closest available station : 5
227
227
Ambulance for event 360 is returning to the station at 101638
Ambulance has arrived on the scene for event 363 at time 101769
Ambulance from event 360 has returned to station 8
Ambulance has arrived on the scene for event 364 at time 101861
Ambulance for event 358 is returning to the station at 101886
Ambulance from event 358 has returned to station 17
Ambulance for event 361 is returning to the station at 102267
Ambulance for event 362 is return

Ambulance has arrived on the scene for event 387 at time 111524
Ambulance for event 385 has arrived at the hospital at time 111571
Ambulance has arrived on the scene for event 389 at time 111750
Ambulance for event 388 has arrived at the hospital at time 111832
Ambulance from event 383 has returned to station 4
Ambulance for event 384 is returning to the station at 112070
calling event id: 390 time: 112203 value: 1561
closest available station : 4
541
541
Ambulance for event 387 has arrived at the hospital at time 112416
Ambulance from event 384 has returned to station 3
Ambulance for event 385 is returning to the station at 112578
Ambulance for event 389 has arrived at the hospital at time 112718
Ambulance for event 388 is returning to the station at 112734
Ambulance has arrived on the scene for event 390 at time 112744
Ambulance from event 385 has returned to station 25
Ambulance from event 388 has returned to station 1
calling event id: 391 time: 113135 value: 1320
closest available

Ambulance has arrived on the scene for event 414 at time 125809
Ambulance for event 408 is returning to the station at 125862
Ambulance for event 409 is returning to the station at 125897
calling event id: 416 time: 125944 value: 1869
closest available station : 25
524
524
Ambulance for event 405 is returning to the station at 126007
Ambulance for event 407 is returning to the station at 126126
Ambulance for event 411 has arrived at the hospital at time 126153
Ambulance has arrived on the scene for event 416 at time 126468
Ambulance for event 410 is returning to the station at 126492
Ambulance from event 408 has returned to station 22
Ambulance from event 409 has returned to station 4
Ambulance from event 405 has returned to station 14
Ambulance from event 407 has returned to station 16
Ambulance for event 412 has arrived at the hospital at time 126632
Ambulance for event 415 has arrived at the hospital at time 126702
Ambulance for event 414 has arrived at the hospital at time 126856
A

calling event id: 439 time: 134045 value: 1319
closest available station : 11
476
476
calling event id: 440 time: 134048 value: 1748
closest available station : 25
482
482
calling event id: 441 time: 134072 value: 1682
closest available station : 14
284
284
Ambulance for event 437 has arrived at the hospital at time 134086
Ambulance from event 432 has returned to station 5
Ambulance from event 426 has returned to station 33
calling event id: 442 time: 134254 value: 1622
closest available station : 17
564
564
Ambulance for event 435 is returning to the station at 134283
Ambulance for event 434 is returning to the station at 134303
Ambulance has arrived on the scene for event 441 at time 134356
Ambulance from event 434 has returned to station 17
Ambulance has arrived on the scene for event 439 at time 134521
Ambulance has arrived on the scene for event 440 at time 134530
calling event id: 443 time: 134665 value: 1568
closest available station : 30
577
577
calling event id: 444 time: 1347

Ambulance from event 452 has returned to station 5
Ambulance for event 456 is returning to the station at 138084
Ambulance from event 451 has returned to station 22
calling event id: 469 time: 138114 value: 1513
closest available station : 30
430
430
Ambulance for event 459 is returning to the station at 138163
Ambulance has arrived on the scene for event 466 at time 138208
Ambulance from event 455 has returned to station 8
calling event id: 470 time: 138333 value: 1314
closest available station : 15
412
412
Ambulance has arrived on the scene for event 468 at time 138401
calling event id: 471 time: 138425 value: 1441
closest available station : 25
601
601
Ambulance for event 458 is returning to the station at 138456
Ambulance for event 464 has arrived at the hospital at time 138502
calling event id: 472 time: 138511 value: 1665
closest available station : 7
488
488
Ambulance for event 462 has arrived at the hospital at time 138528
Ambulance has arrived on the scene for event 469 at tim

Ambulance has arrived on the scene for event 491 at time 143879
calling event id: 493 time: 143931 value: 1009
closest available station : 3
230
230
calling event id: 494 time: 143989 value: 1378
closest available station : 15
597
597
Ambulance has arrived on the scene for event 489 at time 144000
Ambulance from event 481 has returned to station 11
Ambulance has arrived on the scene for event 490 at time 144114
Ambulance for event 483 is returning to the station at 144128
Ambulance has arrived on the scene for event 493 at time 144161
Ambulance from event 479 has returned to station 44
Ambulance for event 482 has arrived at the hospital at time 144183
calling event id: 495 time: 144352 value: 1132
closest available station : 11
563
563
calling event id: 496 time: 144477 value: 1009
closest available station : 8
286
286
Ambulance has arrived on the scene for event 494 at time 144586
Ambulance for event 485 is returning to the station at 144590
Ambulance has arrived on the scene for even

Ambulance for event 502 is returning to the station at 148889
Ambulance for event 506 is returning to the station at 148943
Ambulance for event 511 has arrived at the hospital at time 148963
Ambulance has arrived on the scene for event 514 at time 148998
Ambulance from event 501 has returned to station 4
calling event id: 523 time: 149066 value: 950
closest available station : 19
384
384
Ambulance for event 508 has arrived at the hospital at time 149093
Ambulance for event 510 has arrived at the hospital at time 149115
calling event id: 524 time: 149123 value: 1500
closest available station : 4
489
489
calling event id: 525 time: 149140 value: 950
closest available station : 8
515
515
Ambulance has arrived on the scene for event 522 at time 149172
Ambulance has arrived on the scene for event 520 at time 149195
Ambulance for event 516 has arrived at the hospital at time 149202
Ambulance has arrived on the scene for event 521 at time 149218
Ambulance from event 500 has returned to statio

calling event id: 541 time: 153335 value: 1255
closest available station : 4
643
643
Ambulance for event 533 is returning to the station at 153340
Ambulance has arrived on the scene for event 539 at time 153364
Ambulance has arrived on the scene for event 540 at time 153443
Ambulance for event 532 is returning to the station at 153448
Ambulance for event 536 is returning to the station at 153496
Ambulance from event 536 has returned to station 8
calling event id: 542 time: 153681 value: 2045
closest available station : 17
447
447
calling event id: 543 time: 153693 value: 1254
closest available station : 8
792
792
Ambulance from event 533 has returned to station 15
Ambulance from event 532 has returned to station 17
calling event id: 544 time: 153900 value: 1010
closest available station : 19
639
639
Ambulance for event 538 has arrived at the hospital at time 153908
Ambulance has arrived on the scene for event 541 at time 153978
Ambulance for event 540 has arrived at the hospital at tim

Ambulance from event 553 has returned to station 14
calling event id: 567 time: 159513 value: 948
closest available station : 3
204
204
Ambulance has arrived on the scene for event 566 at time 159540
Ambulance for event 562 has arrived at the hospital at time 159587
Ambulance has arrived on the scene for event 567 at time 159717
Ambulance for event 557 is returning to the station at 159820
Ambulance for event 565 has arrived at the hospital at time 159824
Ambulance for event 558 is returning to the station at 159840
Ambulance for event 561 is returning to the station at 159858
Ambulance for event 560 is returning to the station at 159897
Ambulance for event 559 is returning to the station at 159962
calling event id: 568 time: 160015 value: 1255
closest available station : 5
483
483
Ambulance for event 564 has arrived at the hospital at time 160046
Ambulance from event 559 has returned to station 8
calling event id: 569 time: 160117 value: 1010
closest available station : 8
501
501
call

Ambulance has arrived on the scene for event 589 at time 167295
calling event id: 592 time: 167325 value: 1134
closest available station : 5
539
539
Ambulance for event 586 has arrived at the hospital at time 167422
Ambulance for event 584 is returning to the station at 167424
Ambulance has arrived on the scene for event 591 at time 167462
Ambulance for event 588 has arrived at the hospital at time 167691
Ambulance for event 587 has arrived at the hospital at time 167774
Ambulance has arrived on the scene for event 590 at time 167777
Ambulance from event 582 has returned to station 22
calling event id: 593 time: 167824 value: 834
closest available station : 44
484
484
calling event id: 594 time: 167850 value: 1010
closest available station : 3
347
347
Ambulance has arrived on the scene for event 592 at time 167864
Ambulance from event 584 has returned to station 25
Ambulance for event 589 has arrived at the hospital at time 167965
Ambulance for event 585 is returning to the station at 

Ambulance for event 608 is returning to the station at 174140
calling event id: 619 time: 174184 value: 1498
closest available station : 15
493
493
Ambulance has arrived on the scene for event 617 at time 174306
Ambulance for event 614 has arrived at the hospital at time 174328
Ambulance for event 613 has arrived at the hospital at time 174341
Ambulance from event 604 has returned to station 17
Ambulance from event 607 has returned to station 5
Ambulance has arrived on the scene for event 615 at time 174443
calling event id: 620 time: 174477 value: 828
closest available station : 8
513
513
calling event id: 621 time: 174497 value: 1010
closest available station : 5
333
333
Ambulance for event 616 has arrived at the hospital at time 174677
Ambulance has arrived on the scene for event 619 at time 174677
Ambulance for event 612 is returning to the station at 174751
Ambulance for event 611 is returning to the station at 174751
Ambulance from event 608 has returned to station 22
Ambulance f

Ambulance for event 636 has arrived at the hospital at time 178958
Ambulance has arrived on the scene for event 646 at time 178979
Ambulance from event 623 has returned to station 17
Ambulance has arrived on the scene for event 643 at time 179006
Ambulance has arrived on the scene for event 645 at time 179045
Ambulance for event 639 has arrived at the hospital at time 179137
Ambulance for event 630 is returning to the station at 179269
Ambulance from event 626 has returned to station 1
Ambulance from event 628 has returned to station 11
Ambulance from event 629 has returned to station 4
Ambulance for event 631 is returning to the station at 179515
Ambulance has arrived on the scene for event 647 at time 179530
Ambulance has arrived on the scene for event 644 at time 179532
calling event id: 648 time: 179539 value: 1125
closest available station : 28
408
408
Ambulance for event 638 is returning to the station at 179568
Ambulance from event 630 has returned to station 5
Ambulance for eve

Ambulance from event 655 has returned to station 8
Ambulance for event 661 is returning to the station at 183956
Ambulance for event 659 is returning to the station at 183989
Ambulance has arrived on the scene for event 660 at time 184049
Ambulance has arrived on the scene for event 668 at time 184088
Ambulance has arrived on the scene for event 667 at time 184102
calling event id: 669 time: 184185 value: 641
closest available station : 22
387
387
Ambulance for event 663 is returning to the station at 184224
calling event id: 670 time: 184235 value: 764
closest available station : 8
464
464
Ambulance for event 662 is returning to the station at 184297
Ambulance for event 665 has arrived at the hospital at time 184331
Ambulance from event 658 has returned to station 9
Ambulance from event 659 has returned to station 11
Ambulance for event 664 has arrived at the hospital at time 184479
calling event id: 671 time: 184505 value: 1397
closest available station : 11
1914
1914
Ambulance from 

calling event id: 689 time: 194825 value: 642
closest available station : 22
446
446
Ambulance for event 687 is returning to the station at 194868
Ambulance from event 688 has returned to station 1
Ambulance has arrived on the scene for event 689 at time 195271
calling event id: 690 time: 195299 value: 1379
closest available station : 14
614
614
calling event id: 691 time: 195391 value: 701
closest available station : 9
168
168
Ambulance from event 687 has returned to station 4
Ambulance has arrived on the scene for event 691 at time 195559
Ambulance has arrived on the scene for event 690 at time 195913
Ambulance for event 689 has arrived at the hospital at time 195996
Ambulance for event 690 has arrived at the hospital at time 196645
calling event id: 692 time: 196875 value: 1500
closest available station : 4
489
489
Ambulance for event 689 is returning to the station at 196945
calling event id: 693 time: 197174 value: 886
closest available station : 8
197
197
calling event id: 694 ti

264
264
Ambulance has arrived on the scene for event 715 at time 209589
Ambulance for event 713 is returning to the station at 209591
Ambulance from event 713 has returned to station 17
Ambulance for event 714 has arrived at the hospital at time 209825
Ambulance for event 712 is returning to the station at 209835
Ambulance from event 709 has returned to station 30
Ambulance from event 712 has returned to station 17
Ambulance for event 714 is returning to the station at 210639
Ambulance for event 715 has arrived at the hospital at time 210839
calling event id: 716 time: 210845 value: 1440
closest available station : 4
406
406
Ambulance from event 714 has returned to station 5
Ambulance has arrived on the scene for event 716 at time 211251
Ambulance for event 715 is returning to the station at 211745
calling event id: 717 time: 211823 value: 1197
closest available station : 11
191
191
Ambulance has arrived on the scene for event 717 at time 212014
Ambulance from event 715 has returned to

333
333
calling event id: 744 time: 219809 value: 828
closest available station : 19
328
328
calling event id: 745 time: 219836 value: 1563
closest available station : 17
877
877
Ambulance for event 734 is returning to the station at 219859
calling event id: 746 time: 219864 value: 766
closest available station : 8
368
368
Ambulance has arrived on the scene for event 741 at time 220060
Ambulance for event 735 is returning to the station at 220095
Ambulance for event 736 is returning to the station at 220122
Ambulance has arrived on the scene for event 744 at time 220137
Ambulance has arrived on the scene for event 743 at time 220138
Ambulance from event 734 has returned to station 17
Ambulance has arrived on the scene for event 746 at time 220232
Ambulance from event 735 has returned to station 1
Ambulance has arrived on the scene for event 742 at time 220476
Ambulance for event 740 has arrived at the hospital at time 220480
Ambulance for event 738 is returning to the station at 220487

523
523
Ambulance from event 757 has returned to station 25
Ambulance from event 756 has returned to station 4
calling event id: 769 time: 225289 value: 890
closest available station : 1
466
466
Ambulance has arrived on the scene for event 764 at time 225303
Ambulance for event 763 has arrived at the hospital at time 225320
Ambulance from event 759 has returned to station 34
calling event id: 770 time: 225476 value: 580
closest available station : 22
434
434
Ambulance for event 762 has arrived at the hospital at time 225485
Ambulance for event 765 has arrived at the hospital at time 225656
calling event id: 771 time: 225679 value: 1868
closest available station : 25
348
348
Ambulance has arrived on the scene for event 768 at time 225721
Ambulance has arrived on the scene for event 769 at time 225755
Ambulance for event 766 has arrived at the hospital at time 225792
calling event id: 772 time: 225812 value: 1626
closest available station : 4
578
578
Ambulance for event 764 has arrived a

closest available station : 14
398
398
Ambulance has arrived on the scene for event 794 at time 231142
Ambulance for event 789 has arrived at the hospital at time 231153
Ambulance for event 788 has arrived at the hospital at time 231273
calling event id: 797 time: 231424 value: 577
closest available station : 9
281
281
Ambulance for event 786 has arrived at the hospital at time 231429
Ambulance has arrived on the scene for event 791 at time 231458
Ambulance has arrived on the scene for event 796 at time 231472
Ambulance for event 787 has arrived at the hospital at time 231524
Ambulance for event 784 is returning to the station at 231541
Ambulance has arrived on the scene for event 795 at time 231632
Ambulance has arrived on the scene for event 797 at time 231705
Ambulance has arrived on the scene for event 793 at time 231815
Ambulance for event 792 has arrived at the hospital at time 231895
Ambulance for event 790 has arrived at the hospital at time 231924
Ambulance for event 785 is re

closest available station : 4
254
254
Ambulance has arrived on the scene for event 817 at time 237135
Ambulance from event 805 has returned to station 2
Ambulance has arrived on the scene for event 818 at time 237227
Ambulance has arrived on the scene for event 815 at time 237246
Ambulance has arrived on the scene for event 816 at time 237273
calling event id: 822 time: 237279 value: 1623
closest available station : 25
269
269
Ambulance for event 812 is returning to the station at 237368
Ambulance has arrived on the scene for event 821 at time 237379
Ambulance from event 809 has returned to station 17
Ambulance for event 810 is returning to the station at 237450
Ambulance has arrived on the scene for event 819 at time 237474
calling event id: 823 time: 237479 value: 1314
closest available station : 14
724
724
Ambulance has arrived on the scene for event 822 at time 237548
Ambulance from event 810 has returned to station 8
Ambulance for event 813 is returning to the station at 237628
ca

480
calling event id: 846 time: 241918 value: 519
closest available station : 9
722
722
Ambulance for event 838 is returning to the station at 241934
Ambulance for event 837 has arrived at the hospital at time 241946
Ambulance for event 839 has arrived at the hospital at time 241961
calling event id: 847 time: 241983 value: 392
closest available station : 2
212
212
Ambulance for event 836 is returning to the station at 241990
Ambulance for event 840 has arrived at the hospital at time 242063
Ambulance from event 831 has returned to station 3
Ambulance for event 843 has arrived at the hospital at time 242161
calling event id: 848 time: 242185 value: 1442
closest available station : 30
739
739
Ambulance has arrived on the scene for event 847 at time 242195
Ambulance for event 844 has arrived at the hospital at time 242213
Ambulance for event 841 is returning to the station at 242214
Ambulance for event 832 is returning to the station at 242242
Ambulance from event 833 has returned to sta

closest available station : 11
1055
1055
Ambulance for event 859 is returning to the station at 246613
Ambulance for event 863 is returning to the station at 246624
Ambulance for event 869 has arrived at the hospital at time 246681
calling event id: 871 time: 246708 value: 1682
closest available station : 17
511
511
Ambulance from event 856 has returned to station 17
calling event id: 872 time: 246731 value: 1069
closest available station : 3
494
494
Ambulance for event 850 is returning to the station at 246781
Ambulance for event 868 has arrived at the hospital at time 246821
Ambulance for event 866 has arrived at the hospital at time 246852
Ambulance for event 862 is returning to the station at 246852
Ambulance for event 860 is returning to the station at 246912
calling event id: 873 time: 246958 value: 1247
closest available station : 28
475
475
Ambulance for event 861 is returning to the station at 246986
Ambulance from event 863 has returned to station 1
Ambulance for event 865 is

calling event id: 894 time: 251783 value: 579
closest available station : 22
269
269
calling event id: 895 time: 251800 value: 1687
closest available station : 25
477
477
Ambulance has arrived on the scene for event 891 at time 251912
calling event id: 896 time: 251962 value: 950
closest available station : 3
368
368
calling event id: 897 time: 252000 value: 721
closest available station : 2
300
300
calling event id: 898 time: 252040 value: 1687
closest available station : 14
849
849
Ambulance has arrived on the scene for event 894 at time 252052
Ambulance has arrived on the scene for event 893 at time 252245
Ambulance for event 890 is returning to the station at 252247
Ambulance has arrived on the scene for event 895 at time 252277
Ambulance has arrived on the scene for event 897 at time 252300
calling event id: 899 time: 252300 value: 888
closest available station : 8
318
318
Ambulance has arrived on the scene for event 896 at time 252330
calling event id: 900 time: 252339 value: 113

closest available station : 17
235
235
Ambulance for event 912 is returning to the station at 257742
Ambulance from event 910 has returned to station 15
calling event id: 919 time: 257797 value: 712
closest available station : 43
441
441
Ambulance has arrived on the scene for event 917 at time 257818
Ambulance has arrived on the scene for event 918 at time 257826
Ambulance for event 911 is returning to the station at 258006
calling event id: 920 time: 258007 value: 1135
closest available station : 4
528
528
Ambulance from event 909 has returned to station 30
calling event id: 921 time: 258031 value: 1984
closest available station : 26
369
369
Ambulance for event 916 has arrived at the hospital at time 258104
Ambulance for event 914 is returning to the station at 258116
calling event id: 922 time: 258191 value: 1985
closest available station : 17
272
272
Ambulance has arrived on the scene for event 919 at time 258238
calling event id: 923 time: 258327 value: 762
closest available statio

Ambulance has arrived on the scene for event 943 at time 262748
Ambulance has arrived on the scene for event 947 at time 262761
Ambulance from event 929 has returned to station 26
Ambulance for event 941 has arrived at the hospital at time 262814
Ambulance for event 931 is returning to the station at 262855
calling event id: 948 time: 262908 value: 581
closest available station : 22
628
628
calling event id: 949 time: 262939 value: 1868
closest available station : 4
884
884
Ambulance for event 934 is returning to the station at 262986
Ambulance for event 936 is returning to the station at 263047
Ambulance for event 940 has arrived at the hospital at time 263096
Ambulance from event 928 has returned to station 44
Ambulance for event 935 is returning to the station at 263134
Ambulance for event 946 has arrived at the hospital at time 263146
Ambulance from event 937 has returned to station 3
calling event id: 950 time: 263190 value: 1194
closest available station : 3
467
467
Ambulance for

closest available station : 26
477
477
Ambulance from event 961 has returned to station 11
Ambulance from event 957 has returned to station 19
Ambulance has arrived on the scene for event 967 at time 268597
Ambulance from event 960 has returned to station 8
Ambulance from event 958 has returned to station 26
Ambulance has arrived on the scene for event 968 at time 268754
Ambulance for event 964 is returning to the station at 268757
calling event id: 970 time: 268763 value: 1802
closest available station : 17
224
224
Ambulance for event 965 is returning to the station at 268773
Ambulance for event 966 has arrived at the hospital at time 268802
Ambulance has arrived on the scene for event 969 at time 268804
Ambulance from event 963 has returned to station 41
Ambulance has arrived on the scene for event 970 at time 268987
Ambulance from event 965 has returned to station 22
Ambulance from event 962 has returned to station 44
Ambulance for event 969 has arrived at the hospital at time 26957

236
236
Ambulance has arrived on the scene for event 993 at time 280670
Ambulance for event 991 has arrived at the hospital at time 280803
Ambulance for event 989 is returning to the station at 280843
calling event id: 994 time: 280862 value: 1261
closest available station : 4
388
388
Ambulance for event 990 is returning to the station at 281117
Ambulance from event 989 has returned to station 17
Ambulance from event 987 has returned to station 44
Ambulance for event 992 has arrived at the hospital at time 281232
Ambulance has arrived on the scene for event 994 at time 281250
Ambulance for event 993 has arrived at the hospital at time 281398
Ambulance from event 990 has returned to station 17
calling event id: 995 time: 281510 value: 1574
closest available station : 30
331
331
Ambulance for event 991 is returning to the station at 281758
Ambulance has arrived on the scene for event 995 at time 281841
Ambulance from event 991 has returned to station 17
Ambulance for event 992 is returni

(1000×11 DataFrame
  Row │ id     dispatch_from  waittime  responsetime  scenetime  conveytime  r ⋯
      │ Int64  Int64          Float64   Float64       Float64    Float64     F ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1              1       0.0       2.38333   11.3833      15.3167    ⋯
    2 │     2             31       0.0      10.7667    13.0         18.7167
    3 │     3             25       0.0       7.95      12.2667      17.4333
    4 │     4             33       0.0       5.46667   13.35        18.3833
    5 │     5             14       0.0       6.25      11.25        15.9167    ⋯
    6 │     6             22       0.0       3.7       16.25        16.2833
    7 │     7             16       0.0      15.45      15.6167      18.5333
    8 │     8              8       0.0       2.55      19.8333      12.4833
    9 │     9             22       0.0       6.45       9.08333     15.7667    ⋯
   10 │    10              8       0.0 

In [60]:
println(mean(df[!,"responsetime"]))
println(mean(df[!,"scenetime"]))
println(mean(df[!,"conveytime"]))
println(mean(df[!,"returntime"]))

8.145316666666668
16.25061666666667
15.4032
9.455366666666668


## Temp results journal 

println(mean(df["responsetime"]))
println(mean(df["scenetime"]))
println(mean(df["conveytime"]))
println(mean(df["returntime"]))
println(mean(df[!,"returntime"]))
#### lognromal 3.08
9.813166666666667 <br>
9.134116666666667 <br>
24.4176 <br>
8.908 <br>
#### lognromal 3.65
10.3443 <br>
16.000433333333337 <br>
24.4176 <br>
9.090300000000001 <br>
#### adjust travel time by time_bin weights
10.466350000000002 <br>
16.000433333333334 <br>
24.4176 <br>
9.073 <br>
#### global
7.7477 <br>
16.250616666666666 <br>
24.4176 <br>
9.114916666666668 <br>
#### log-log 3200 grid UNIFORM 44 WHOOPS
5.3978 <br>
16.000433333333337 <br>
15.4032 <br>
8.835816666666668 <br>
#### 3200 grid UNIFORM 44  WHOOPS
6.387766666666668 <br>
16.000433333333334 <br>
15.4032 <br>
8.805600000000002 <br>
#### log-log 3200 grid stochastic
5.6897166666666665 <br>
16.000433333333337 <br>
15.4032 <br>
9.427483333333335 <br>
#### 3200 grid stochastic
8.145316666666668
16.25061666666667
15.4032
9.455366666666668


In [27]:
"""batch call simulation"""
# batches = 12
# test_calls = CSV.File("../test/austin-data/austin_test_calls_update1.csv")|> DataFrame 
# #update 1: remove negative indices
# using DelimitedFiles

# for i = 1:12
#     """ simulation parameters """
#     p = DeploymentProblem(
#               hourly_calls,
#               adjacent_nbhd,
#               coverage,
#               namb = namb,
#               train_filter = (hourly_calls[!,:year] .== 2019) .* (hourly_calls[!,:month] .<= 3)
#               )
#     test_calls = CSV.File("../test/austin-data/austin_test_calls_update1.csv")|> DataFrame 
#     #update 1: remove negative indices
#     start_ind = (i-1)*1000+1
#     end_ind = (i)*1000
#     println("running $namb ambuances & $start_ind to $end_ind calls")
#     test_calls_batch = test_calls[start_ind:end_ind,:]

#     # distribution of ambulances across all of the stations
#     x = amb_deployment[model_dict[model_name]][namb]

#     #establishes the dispatch problem
#     problem = DispatchProblem(test_calls_batch, hospitals, stations, p.coverage, x, turnaround=turnaround)

#     #establishes the dispatch model
#     dispatch = ClosestDispatch(p, problem)

#     #@time df, guiArray = simulate_events!(problem, dispatch)
#     df, guiArray = simulate_events!(problem, dispatch)
#     print(mean(df["responsetime"]))
    
#     filename = string(model_name,string(namb),"calls",string(ncalls),"batch",string(i),".csv")
#     open(filename, "w") do filename
#         writedlm(filename,  df["responsetime"])
#     end
#     print(filename)
# end

"batch call simulation"

In [28]:
"""all models simulation"""
# turnaround = Distributions.LogNormal(3.65, 0.3)
# lambda = 0
# local_path = ""
# model_name = "Stochastic"
# namb = 40
# ncalls = 1000
# #using AmbulanceDeployment

# pwd()

# cd(currentpath * "/test/")
# pwd()

# #using AmbulanceDeployment
# using Dates

# #include("..//src//dispatch/closestdispatch.jl")
# include("..//src//problem.jl")
# #include("..//src//evaluate.jl") #if you include, you need to include JuMP
# #include("..//src//simulate.jl")

# hospitals = CSV.File(string(local_path,"../test/austin-data/hospitals.csv")) |> DataFrame
# stations = CSV.File(string(local_path,"../test/austin-data/stations.csv")) |> DataFrame
# hourly_calls = CSV.File("../test/austin-data/Full_WeekdayCalls.csv") |> DataFrame
# adjacent_nbhd = CSV.File("../test/austin-data/adjacent_nbhd.csv") |> DataFrame
# coverage = CSV.read("../test/austin-data/coverage_real.csv", DataFrame, header=false)
# coverage = convert(Array{Bool, 2}, coverage[:, :])
# incidents = CSV.File("../test/austin-data/austin_incidents.csv") |> DataFrame
# solverstats = JLD.load("../src/outputs/austin_team_stats_1_03.jld")
# amb_deployment = solverstats["amb_deployment"]
# model_dict = Dict{String, Symbol}("Stochastic"=>:Stochastic, "Robust01"=>:Robust01, "Robust005"=>:Robust005, "Robust001"=>:Robust001, "Robust0001"=>:Robust0001,
#           "Robust00001"=>:Robust00001, "MEXCLP"=>:MEXCLP, "MALP"=>:MALP)


# p = DeploymentProblem(
#           hourly_calls,
#           adjacent_nbhd,
#           coverage,
#           namb = namb,
#           train_filter = (hourly_calls[!,:year] .== 2019) .* (hourly_calls[!,:month] .<= 3)
#           )
# test_calls = CSV.File("../test/austin-data/austin_test_calls_update1.csv")|> DataFrame 
# #update 1: remove negative indices
# test_calls = test_calls[1:ncalls,:]
# println("running $namb ambuances & $ncalls calls")

# # distribution of ambulances across all of the stations
# x = amb_deployment[model_dict[model_name]][namb]

# #establishes the dispatch problem
# problem = DispatchProblem(test_calls, hospitals, stations, p.coverage, x, turnaround=turnaround)

# #establishes the dispatch model
# dispatch = ClosestDispatch(p, problem)

# Random.seed!(1234)

"all models simulation"

In [29]:
array = filter(x->typeof(x)==gui_event,guiArray)
responded_array = filter(x->x.event_type == "call responded", array)

1000-element Array{Any,1}:
 gui_event("call responded", 1012, 1, 1, 1, -1, 0, DateTime("2021-07-19T12:39:09.102"))
 gui_event("call responded", 220, 31, 2, 31, -1, 0, DateTime("2021-07-19T12:39:09.196"))
 gui_event("call responded", 1687, 36, 3, 36, -1, 0, DateTime("2021-07-19T12:39:09.258"))
 gui_event("call responded", 1536, 33, 4, 33, -1, 0, DateTime("2021-07-19T12:39:09.304"))
 gui_event("call responded", 1621, 14, 5, 14, -1, 0, DateTime("2021-07-19T12:39:09.397"))
 gui_event("call responded", 578, 22, 6, 22, -1, 0, DateTime("2021-07-19T12:39:09.49"))
 gui_event("call responded", 1169, 16, 7, 16, -1, 0, DateTime("2021-07-19T12:39:09.505"))
 gui_event("call responded", 825, 8, 8, 8, -1, 0, DateTime("2021-07-19T12:39:09.721"))
 gui_event("call responded", 641, 9, 9, 9, -1, 0, DateTime("2021-07-19T12:39:09.8"))
 gui_event("call responded", 825, 3, 10, 3, -1, 0, DateTime("2021-07-19T12:39:09.892"))
 gui_event("call responded", 765, 35, 11, 35, -1, 0, DateTime("2021-07-19T12:39:09.985")

In [61]:
using DelimitedFiles
filename = string("visualization/time_adjust/",model_name,string(namb),"calls",string(ncalls),"_3200grid",".csv")
open(filename, "w") do filename
    writedlm(filename,  df[!,"responsetime"])
end
print(filename)

visualization/time_adjust/Stochastic40calls1000_3200grid.csv

In [48]:
#using CSV
#filename = string(joshpath,"test/",model_name,string(namb),"calls",string(ncalls),".csv")
#CSV.write(filename, df, delim=",")